.github/workflows/ghcr-publish-release.yaml

```diff
@@ -21,26 +27,41 @@ jobs:
     strategy:
       fail-fast: true
     steps:
-      - name: Check out the repo
+      - name: Check out the repo current push
+        if: ${{ inputs.tag == '' }}
         uses: actions/checkout@v4
+      - name: Check out the repo at specified tag
+        if: ${{ inputs.tag != '' }}
+        uses: actions/checkout@v4
+        with:
+          ref: ${{ inputs.tag }}
       - name: Log in to registry
         uses: docker/login-action@f4ef78c080cd8ba55a85445d5b36e214a81df20a
         with:
           registry: ghcr.io
           username: ${{ github.actor }}
           password: ${{ github.token }}
       - name: Extract metadata (tags, labels) for Docker
+        if: ${{ inputs.tag == '' }}
         id: meta
         uses: docker/metadata-action@9ec57ed1fcdbf14dcef7dfbe97b2010124a938b7
         with:
           images: ${{ env.IMAGE }}
           tags: |
             type=semver,pattern={{version}}
+      - name: Extract passed tag metadata (tags, labels) for Docker
+        if: ${{ inputs.tag != '' }}
+        id: meta-value
+        uses: docker/metadata-action@9ec57ed1fcdbf14dcef7dfbe97b2010124a938b7
+        with:
+          images: ${{ env.IMAGE }}
+          tags: |
+            type=semver,pattern={{version}},value=${{ inputs.tag }}
```

Comment:
Using some environment variables and expressions, we could avoid the duplicate steps. (In the environment variables, you can use the inputs from the workflow)